### PR TITLE
Update testing environment to more recent versions

### PR DESCRIPTION
**Pull request by nbren12**
It's been a while since this repo saw any love, and many dep versions were out of date. I went down this rabbit hole after noticing that the new fsspec broke a highly used test fixture that unfortunately depended on private fsspec APIs (MockFileSystem._strip_protocol)` which had changed. Note we use the newer fsspec in practice in fv3net so this issue is isolated to the test suite.

I removed this mock so that we don't need to maintain it any longer and retargetd at the `MemoryFileSystem` class which it inherited from. 

I also had to change the cache-directory to support URLS like `file:///some/path`, which the `MemoryFileSystem` seems to make now. This bug would have occured even if I had repaired the mock gcsfs object.

*Debrief*: overall, I'm worried about technical debt in this test suite.  It took me 6+ hours to do this PR, all to upgrade to a version of fsspec that we know works well with fv3config from extensive production testing.  It definitely demonstrates the downsides of using mocks of complex objects like fsspec, depending on private external apis in those mocks, and then widely reusing those mock. If it happens again, I think we may want to be pretty aggressive with deleting/consolidating tests.

I think it's mostly historical baggage at this point since fv3config actually has a rather clean design where 1) configs are compiled into a list of assets and 2) then assets are written to disk. The tests could target (1) and (2) separately so we could test the config handling separately from the I/O/paths/etc.

### CODE REVIEW

File: fv3config/filesystem.py [hunk 1]
```diff
@@ -145,11 +146,22 @@ def put_file(source_filename, dest_filename):
...
+    path = pathlib.Path(path_str)
+    cache_dir = pathlib.Path(caching.get_internal_cache_dir()).absolute()
+
+    if path.is_absolute():
+        path_no_root = path.relative_to(path.root)
+        rel_or_abs = "ABS"
+    else:
+        path_no_root = path
+        rel_or_abs = "REL"
+    path_in_cache = cache_dir / rel_or_abs / prefix / path_no_root
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply fv3config/filesystem.py:163]: After our discussion, can you rename ABS and REL to be lowercase? then they should read more like paths (to me at least). You could also rename the variable.

**Review by mcgibbon — APPROVED** [review]
Approved with suggestions. Thanks!

File: fv3config/filesystem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (nbren12) [reply fv3config/filesystem.py]: sure. i  pushed this.

File: tests/test_write_run_directory.py [hunk 2]
```diff
@@ -62,84 +62,13 @@ def test_set_then_get_archive_dir(self):
...
-            fv3config.write_run_directory(config, rundir)
-        cache_filename = os.path.join(
-            fv3config.get_cache_dir(),
-            "fv3config-cache/gs/vcm-fv3config/config/diag_table/default/v1.0/diag_table",
-        )
-        assert os.path.isfile(cache_filename)
-        modification_time = os.path.getmtime(cache_filename)
-        with tempfile.TemporaryDirectory() as rundir:
-            fv3config.write_run_directory(config, rundir)
-        assert os.path.getmtime(cache_filename) == modification_time
```
[the remark this answers is not in the document]
Comment (nbren12) [reply tests/test_write_run_directory.py:142]: we discussed it being ok to lose this coverage and possibly restore it in the form of a more closely target unit test of `_get_file`.

**Pull request merged**